misc/python/materialize/mzcompose/composition.py

```diff
@@ -775,6 +775,10 @@ def down(
         if sanity_restart_mz:
             self.sanity_restart_mz()
 
+        logs = self.invoke("logs", "--no-color", capture=True).stdout
+        with open(MZ_ROOT / "services.log", "a") as f:
```

Comment:
I think there is a some magic incantation where you can redirect stdout to `services.log` with a `with` statement and then the `self.invoke` can be put inside the `with` block. Maybe even without `capture=True`, thus removing all buffering -- `docker logs` can be arbitrarily large so ideally at no point it should be loaded in memory.